omod-2.0/src/test/java/org/openmrs/module/webservices/rest/web/v1_0/controller/openmrs2_0/NameTemplateController2_0Test.java

```diff
@@ -9,64 +9,128 @@
+	
+	private static final String ALTERNATE_NAMETEMPLATE = "long";
+	
+	private static final String ALTERNATE_NAMETEMPLATE_RESOURCE = "nameTemplateLong.json";
+	
+	private static final String UNKNOWN_NAMETEMPLATE = "foo";
+	
+	private static final int NAME_TEMPLATES_COUNT = 5;
+	
+	private String originalNametemplate = null;
+	
+	public String getURI() {
+		return "nametemplate";
+	}
+	
+	@Before
+	public void before() throws Exception {
+		// save a backup of the current global property value for name layout,
+		// before setting a specific name layout value for the tests.
+		AdministrationService service = Context.getAdministrationService();
+		originalNametemplate = service.getGlobalProperty(OpenmrsConstants.GLOBAL_PROPERTY_LAYOUT_NAME_FORMAT);
+		service.setGlobalProperty(OpenmrsConstants.GLOBAL_PROPERTY_LAYOUT_NAME_FORMAT, DEFAULT_NAMETEMPLATE);
+	}
+	
+	@After
+	public void after() throws Exception {
+		if (!StringUtils.isEmpty(originalNametemplate)) {
+			// restore the backed up name layout global property value.
+			AdministrationService service = Context.getAdministrationService();
+			service.setGlobalProperty(OpenmrsConstants.GLOBAL_PROPERTY_LAYOUT_NAME_FORMAT, originalNametemplate);
+		}
+	}
+	
+	@Test
+	public void shouldGetAll() throws Exception {
+		MockHttpServletRequest req = newGetRequest(getURI());
+		SimpleObject result = deserialize(handle(req));
+		
+		SimpleObject defaultNameTemplateResource = parseNameTemplateResource(DEFAULT_NAMETEMPLATE_RESOURCE);
+		SimpleObject alternateNameTemplateResource = parseNameTemplateResource(ALTERNATE_NAMETEMPLATE_RESOURCE);
+		
+		Assert.assertThat(result.containsKey("results"), is(true));
+		Assert.assertThat(result.get("results"), iterableWithSize(NAME_TEMPLATES_COUNT));
+		Assert.assertThat(result.get("results"), hasItem(defaultNameTemplateResource));
+		Assert.assertThat(result.get("results"), hasItem(alternateNameTemplateResource));
+	}
+	
+	@Test
+	public void shouldGetGlobalNameTemplate() throws Exception {
+		MockHttpServletRequest req = newGetRequest(getURI() + "/" + URL_GLOBAL_NAMETEMPLATE);
+		SimpleObject result = deserialize(handle(req));
+		
+		SimpleObject expectedResult = withResourceVersion(parseNameTemplateResource(DEFAULT_NAMETEMPLATE_RESOURCE));
+		
+		Assert.assertThat(result, is(expectedResult));
+	}
+	
+	@Test
+	public void shouldGetNameTemplateByCodename() throws Exception {
+		MockHttpServletRequest req = newGetRequest(getURI() + "/" + ALTERNATE_NAMETEMPLATE);
+		SimpleObject result = deserialize(handle(req));
+		
+		SimpleObject expectedResult = withResourceVersion(parseNameTemplateResource(ALTERNATE_NAMETEMPLATE_RESOURCE));
+		
+		Assert.assertThat(result, is(expectedResult));
+	}
+	
+	@Test
+	public void shouldReturnNotFoundForUnknownCodename() throws Exception {
+		MockHttpServletRequest req = newGetRequest(getURI() + "/" + UNKNOWN_NAMETEMPLATE);
+		MockHttpServletResponse result = null;
+		ObjectNotFoundException handledException = null;
```

Comment:
What do you think about using `org.junit.rules.ExpectedException`?

Reply:
Good idea. Done.